asserts/snapasserts/validation_sets.go

```diff
@@ -463,114 +511,253 @@ func (v *ValidationSets) Conflict() error {
+// target to be uninstalled.
+func checkConstraints(cs constraints, installedRev snap.Revision) (
+	// invalid validation set keys
+	invalid []string,
+	// required revision (could be unspecified) -> invalid validation set keys
+	missing map[snap.Revision][]string,
+	// required revision (should never be unspecified) -> invalid validation set keys
+	wrongrev map[snap.Revision][]string,
+	err error,
+) {
+	invalidSet := make(map[string]bool)
+	missingSets := make(map[snap.Revision]map[string]bool)
+	wrongrevSets := make(map[snap.Revision]map[string]bool)
+
+	for requiredRev, revCstr := range cs.revisions {
+		for _, rc := range revCstr {
+			conformity := checkConstraintConformity(installedRev, requiredRev, cs.presence)
+			switch conformity {
+			case constraintConformityInvalid:
+				invalidSet[rc.validationSetKey] = true
+			case constraintConformityMissing:
+				if missingSets[requiredRev] == nil {
+					missingSets[requiredRev] = make(map[string]bool)
 				}
-
-				switch {
-				case !isInstalled && (cstrs.presence == asserts.PresenceOptional || cstrs.presence == asserts.PresenceInvalid):
-					// not installed, but optional or not required
-				case isInstalled && cstrs.presence == asserts.PresenceInvalid:
-					// installed but not expected to be present
-					if invalid[rc.Name] == nil {
-						invalid[rc.Name] = make(map[string]bool)
-					}
-					invalid[rc.Name][rc.validationSetKey] = true
-				case isInstalled:
-					// presence is either optional or required
-					if rev != unspecifiedRevision && rev != sn.(*InstalledSnap).Revision {
-						// expected a different revision
-						if wrongrev[rc.Name] == nil {
-							wrongrev[rc.Name] = make(map[snap.Revision]map[string]bool)
-						}
-						if wrongrev[rc.Name][rev] == nil {
-							wrongrev[rc.Name][rev] = make(map[string]bool)
-						}
-						wrongrev[rc.Name][rev][rc.validationSetKey] = true
-					}
-				default:
-					// not installed but required.
-					// note, not checking ignoreValidation here because it's not a viable scenario (it's not
-					// possible to have enforced validation set while not having the required snap at all - it
-					// is only possible to have it with a wrong revision, or installed while invalid, in both
-					// cases through --ignore-validation flag).
-					if missing[rc.Name] == nil {
-						missing[rc.Name] = make(map[snap.Revision]map[string]bool)
-					}
-					if missing[rc.Name][rev] == nil {
-						missing[rc.Name][rev] = make(map[string]bool)
-					}
-					missing[rc.Name][rev][rc.validationSetKey] = true
+				missingSets[requiredRev][rc.validationSetKey] = true
+			case constraintConformityWrongRevision:
+				if wrongrevSets[requiredRev] == nil {
+					wrongrevSets[requiredRev] = make(map[string]bool)
 				}
+				wrongrevSets[requiredRev][rc.validationSetKey] = true
+			case constraintConformityValid:
+				continue
+			default:
+				return nil, nil, nil, fmt.Errorf("internal error: unknown conformity %d", conformity)
```

Comment:
This case is really only here because go doesn't have any sort of exhaustive matching on enums, shouldn't be possible unless the programmer does something really wrong.